codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -265,18 +266,21 @@ public Http2Stream resetSent() {
         }
 
         @Override
-        public final Object setProperty(Object key, Object value) {
-            return data.put(key, value);
+        public final <T> T setProperty(Class<T> key, T value) {
+            int index = propertyKeyRegistry.indexOf(checkNotNull(key, "key"));
+            return properties.add(index, checkNotNull(value, "value"));
         }
 
         @Override
-        public final <V> V getProperty(Object key) {
-            return data.get(key);
+        public final <T> T getProperty(Class<T> key) {
```

Comment:
It seems like the relationship between key class type and property value type may lead to collisions and unintended overlap.  For example `EmbeddedChannel` is now used in `CompressorHttp2ConnectionEncoder` as the key (and thus return type) for its internal state.  However the `EmbeddedChannel` type is visible outside of the `CompressorHttp2ConnectionEncoder` class and it is possible that someone else will also use this as a key.  If this happens then we will get collisions and unexpected behavior.  It is also limiting in general because what if a user wants to store more than 1 of any given type?